Work in progress

davidlehn opened the pull request.

davidlehn committed:
Update test and benchmark.

- Use mocha and benchmark packages.
- Move tests to test/.
- Move benchmarks to benchmark/

davidlehn committed:
Use unsigned int api.

davidlehn committed:
Check difficulty in verify.

davidlehn committed:
Add tests.

davidlehn committed:
Add travis support.

davidlehn committed:
Fix typo.

davidlehn committed:
Tests, benchmark, difficulty, misc.

- Add more tests.
- Add difficulty test.
- Add nonce controls.
- Add common tests for use by solver engines.
- Add graphSize=30 test vector.

davidlehn committed:
Expose constants API.

davidlehn committed:
Add histogram display tool.

davidlehn committed:
Update benchmarking.

- Add per-engine testers so each engine can run same benchmarks.

davidlehn committed:
Improve solver.

- Accept engine options.
- Improve looping through potential nonces and solutions.

davidlehn committed:
Add difficulty constants and test support.

davidlehn committed:
Test updates.

- Skip long test.
- Add difficuly test.

davidlehn committed:
Fix test.

davidlehn committed:
Add initial chained verify/solve.

dlongley commented on benchmark/nonces.js:
```diff
@@ -0,0 +1,91 @@
+ *
+ * Copyright (c) 2018 Digital Bazaar, Inc.
+ *
+ * MIT License
```
BSD

dlongley commented on benchmark/verify.js:
```diff
@@ -0,0 +1,37 @@
+const Long = require('long');
+
+const input = Buffer.allocUnsafe(0);
+const edges = '21ce 5240 d304 d34c f756 faf5 16c9f 1a349 1d3d9 2096a 22036 2589b 2e2ed 2eb40 2fb3c 376fd 37740 393c0 3ad29 3cf04 3f365 41fe2 43a29 454eb 4cf13 4d12c 535ed 57d03 60e81 68fd1 6902f 69408 6c2f1 728c8 73e0e 76589 7a037 7adcb 7c4b8 7d746 7eae0 7fe67'
```
eslint will complain about this

dlongley commented on lib/helpers.js:
```diff
@@ -23,6 +23,28 @@ function createHeader({input, nonce}) {
+ */
+api.checkDifficulty = ({edges, difficulty}) => {
+  // FIXME: check if sorted, always sort, or assume edges are sorted?
+  //const sortedEdges = Array.from(edges).sort();
```
I vote always sort.

dlongley commented on lib/helpers.js:
```diff
@@ -23,6 +23,28 @@ function createHeader({input, nonce}) {
+  for(let i = 0; i < sortedEdges.length; ++i) {
+    edgesBuf.writeUInt32LE(sortedEdges[i], i * 4);
+  }
+  const edgesHash = chloride.crypto_generichash(edgesBuf, constants.HASH_SIZE);
```
Can we get blake2b from another lib? `chloride` isn't very actively maintained and we're looking to drop it elsewhere and replace with [sodium-universal](https://github.com/sodium-friends/sodium-universal) or [sodium-native](https://github.com/sodium-friends/sodium-native).

dlongley commented on lib/verify.js:
```diff
@@ -156,7 +158,78 @@ api.verify = async ({
+    throw new TypeError('Chained "solution" must be an array.');
+  }
+  if(!(Number.isInteger(minLinks) && minLinks >= 1)) {
+    throw new TypeError('"minLinks" must be an unsigned integer > 1.');
```
should this and elsewhere read `>= 1` instead of `> 1`?

dlongley commented on lib/verify.js:
```diff
@@ -156,7 +158,78 @@ api.verify = async ({
+    throw new TypeError(`Solution must have at least ${minLinks} links`);
+  }
+  if(solution.length > maxLinks) {
+    throw new TypeError(`Solution must not have more than ${maxLinks} links`);
```
Add periods to error messages.

dlongley reviewed (approved):
LGTM other than some minor nits -- please address before merging. Also, we should add our default `eslint.rc` file and address any issues it flags.

dlongley commented on lib/solve.js:
```diff
@@ -33,25 +39,126 @@ api.solve = async ({
+    // FIXME: document format
+    const hash = crypto.createHash('sha256');
+    hash.update(Int32Array.from(s.nonce));
+    hash.update(Int32Array.from(s.edges));
```
My understanding is that `Int32Array.from` does endianness based on the underlying architecture ... which won't work here. We need to make sure it's always little endian... also, Uint32s.

dlongley replied:
Something like this might work as a helper function:

```js
function toUint32Buffer(x) {
  const b = Buffer.allocUnsafe(x.length * 4);
  x.forEach((v, i) => b.writeUInt32LE(v, i * 4));
  return b;
}
```

davidlehn committed:
Fix license comment.

davidlehn committed:
Add SipHash type support.

davidlehn committed:
Add eslint support.

davidlehn committed:
Fix lint issues.

davidlehn committed:
Fix edge generation.

davidlehn committed:
Sort edges for difficulty check.

davidlehn committed:
Switch from chloride to sodium-universal.

davidlehn committed:
Clean up error messages.

davidlehn committed:
More SipHash type support and tests.

davidlehn committed:
Sort numerically.

davidlehn committed:
Chaining updates.

- Rename "links" to "chainLength".
- Add solution hashing helper.
- Add solution edge length into solution hash.
- Add array of starting nonces (helps with testing).
- Add max nonce checks per chain.
- Update tests.

davidlehn committed:
Add cuckatoo support.

- Add "variant" option to select cuckoo/cuckatoo.
- Update JS verifier to handle cuckatoo variant.
- Update tests.
- Disable engine cuckoo tests that require bindings.
  - Current engine doesn't build cuckoo bindings.
  - TODO is to let engine skip tests it doesn't support.
